typescript/src/wallet.ts

```diff
@@ -209,3 +217,117 @@ type DkgResult = {
+    witnessAddress: boolean
+  ): Promise<UnspentTransactionOutput | undefined> => {
+    // Build the wallet Bitcoin address based on its public key hash.
+    const walletAddress = encodeToBitcoinAddress(
+      walletPublicKeyHash.toString(),
+      witnessAddress,
+      bitcoinNetwork
+    )
+
+    // Get the wallet transaction history. The wallet main UTXO registered in the
+    // Bridge almost always comes from the latest BTC transaction made by the wallet.
+    // However, there may be cases where the BTC transaction was made but their
+    // SPV proof is not yet submitted to the Bridge thus the registered main UTXO
+    // points to the second last BTC transaction. In theory, such a gap between
+    // the actual latest BTC transaction and the registered main UTXO in the
+    // Bridge may be even wider. The exact behavior is a wallet implementation
+    // detail and not a protocol invariant so, it may be subject of changes.
+    // To cover the worst possible cases, we always take the five latest
+    // transactions made by the wallet for consideration.
+    const walletTransactions = await bitcoinClient.getTransactionHistory(
+      walletAddress,
+      5
+    )
+
+    // Get the wallet script based on the wallet address. This is required
+    // to find transaction outputs that lock funds on the wallet.
+    const walletScript = createOutputScriptFromAddress(walletAddress)
+    const isWalletOutput = (output: TransactionOutput) =>
+      walletScript.equals(output.scriptPubKey)
+
+    // Start iterating from the latest transaction as the chance it matches
+    // the wallet main UTXO is the highest.
+    for (let i = walletTransactions.length - 1; i >= 0; i--) {
+      const walletTransaction = walletTransactions[i]
+
+      // Find the output that locks the funds on the wallet. Only such an output
+      // can be a wallet main UTXO.
+      const outputIndex = walletTransaction.outputs.findIndex(isWalletOutput)
+
+      // Should never happen as all transactions come from wallet history. Just
+      // in case check whether the wallet output was actually found.
+      if (outputIndex < 0) {
+        continue
+      }
+
+      // Build a candidate UTXO instance based on the detected output.
+      const utxo: UnspentTransactionOutput = {
+        transactionHash: walletTransaction.transactionHash,
+        outputIndex: outputIndex,
+        value: walletTransaction.outputs[outputIndex].value,
+      }
+
+      // Check whether the candidate UTXO hash matches the main UTXO hash stored
+      // on the Bridge.
+      if (mainUtxoHash.equals(bridge.buildUtxoHash(utxo))) {
+        return utxo
+      }
+    }
+
+    return undefined
+  }
+
+  // The most common case is that the wallet uses a witness address for all
+  // operations. Try to determine the main UTXO for that address first as the
+  // chance for success is the highest here.
+  const mainUtxo = await determine(true)
+  if (mainUtxo) {
+    return mainUtxo
+  }
```

Comment:
We can simplify `rerturn mainUtxo ?? determine(false)`

Reply:
See 7b0b9bdb